x/evm/keeper/msg_assigner.go

```diff
@@ -3,266 +3,324 @@ package keeper
+		return window
+	})
+
+	if !pdw.assert() {
+		// If we can't assert the metadata, we can't rank the validators.
+		return scoreSnapshot{}, errors.New("performance data window assertion failed")
 	}
 
-	assignableValidators := filterAssignableValidators(currentSnapshot.Validators, chainID, req)
-	if len(assignableValidators) == 0 {
-		return "", errors.New("no assignable validators for message")
+	for valAddr, info := range validatorsInfos {
+		scores := map[string]math.LegacyDec{
+			"fee":           scoreValue(*pdw.maxFee, *pdw.minFee, info.Fee, true),
+			"uptime":        scoreValue(*pdw.maxUptime, *pdw.minUptime, info.Uptime, false),
+			"successRate":   scoreValue(*pdw.maxSuccessRate, *pdw.minSuccessRate, info.SuccessRate, false),
+			"executionTime": scoreValue(*pdw.maxExecutionTime, *pdw.minExecutionTime, info.ExecutionTime, true),
+			"featureSet":    scoreValue(*pdw.maxFeatureSet, *pdw.minFeatureSet, info.FeatureSet, false),
+		}
+
+		score := (scores["fee"].Mul(relayWeights.Fee)).Add(
+			(scores["uptime"].Mul(relayWeights.Uptime))).Add(
+			(scores["successRate"].Mul(relayWeights.SuccessRate))).Add(
+			(scores["executionTime"].Mul(relayWeights.ExecutionTime))).Add(
+			(scores["featureSet"].Mul(relayWeights.FeatureSet)))
+
+		ranked = append(ranked, validatorScore{address: valAddr, score: score})
 	}
 
-	validatorsInfos := buildValidatorsInfos(assignableValidators)
+	// Sort by score
+	slices.SortStableFunc(ranked, func(a, b validatorScore) int {
+		if a.score.GT(b.score) {
+			return 1
+		} else if a.score.LT(b.score) {
+			return -1
+		}
+		return 0
+	})
+	// Slice is sorted ascending, but we're interested in high scorers.
+	slices.Reverse(ranked)
+	return scoreSnapshot{
+		scores:      ranked,
+		blockHeight: sdkCtx.BlockHeight(),
+	}, nil
+}
 
-	if weights == nil {
-		weights = &types.RelayWeights{
-			ExecutionTime: "1.0",
-			SuccessRate:   "1.0",
-			Uptime:        "1.0",
-			Fee:           "1.0",
-			FeatureSet:    "1.0",
+// filterValidatorsForJob takes a sorted list of validators and filters them based on the job requirements.
+func filterValidatorsForJob(scoreSnapshot scoreSnapshot, valLUT map[string]valsettypes.Validator, chainID string, req *xchain.JobRequirements) []validatorScore {
+	return slice.Filter(scoreSnapshot.scores, func(score validatorScore) bool {
+		// Snapshot validator not found in lookup.
+		// This shouldn't really happen ever, but better to catch it just in case.
+		val, ok := valLUT[score.address]
+		if !ok {
+			return false
 		}
-	}
-	relayWeights := weights.Float64Values()
+		for _, v := range val.ExternalChainInfos {
+			// Has support for chain. Should always be the case.
+			if v.ChainReferenceID != chainID {
+				continue
+			}
 
-	return pickValidator(ctx, validatorsInfos, relayWeights), nil
+			// Is MEV support required?
+			if req == nil || !req.EnforceMEVRelay {
+				return true
+			}
+
+			for _, t := range v.Traits {
+				if req.EnforceMEVRelay && t == valsettypes.PIGEON_TRAIT_MEV {
```

Comment:
You're right. I think I kept this as explicit as possible to be clear about what's going on. I'd probably leave it like this unless you feel strongly otherwise.